Introduce deref cmd

amir73il opened the pull request:
**deref** - copy changes from `upperdir` to `uppernew` while unfolding redirect directories and metacopy regular files, so that new upperdir is compatible with legacy overlayfs driver.

amir73il committed:
Add support for command diff -b|--brief

Comforms to the output format of 'diff -qr --no-dereference'.

Signed-off-by: Amir Goldstein <amir73il@gmail.com>

amir73il committed:
Add makefile for running test cases

Expected output of 'overlay diff' is comitted as diff.saved.
Expected output of 'overlay diff -v' is comitted as verbose.saved.
Expected output of 'overlay diff -b' is generated with 'diff -qr'.

Signed-off-by: Amir Goldstein <amir73il@gmail.com>

amir73il committed:
Add awareness to redirect dir

diff treats redirect dir as opaque -
- lower files under redirect dir listed as deleted
- upper dir listed as added

vaccum also treats redirect dir as opaque -
- it must not delete redirect upper dir becaue we do not
  know how to follow redirect to lower

merge with redirect dir is complicated and not supported -
- we abort merge when detecting a redirect dir to avoid
  deleting files under the redirected dir

Signed-off-by: Amir Goldstein <amir73il@gmail.com>

amir73il committed:
Make verbose a global var and callbacks funcs static

Signed-off-by: Amir Goldstein <amir73il@gmail.com>

amir73il committed:
Add awareness for upper metacopy

Data of metacopy upper is identical to lower by definition
unless upper is redirected, so diff and vacuum work.

merge is more complicated, so it is aborted if an upper metacopy
is found.

Signed-off-by: Amir Goldstein <amir73il@gmail.com>

amir73il committed:
Add a metacopy file to diff tests

It is not a real metacopy because it is missing the origin xattr,
but for the diff tests it is sufficient.

In the test data set, the victim file was unmodified, only permission
bits changed, so diff results remain the same.

Signed-off-by: Amir Goldstein <amir73il@gmail.com>

amir73il committed:
Add renamed (redirect) dir to diff tests

"fake" that lower removed_dir was actually renamed and not removed,
so all its children can be found under renamed_dir.

The diff output results of diff and diff -b have an extra line for
added directory /renamed_dir.

The diff output of diff -v should have the children of renamed_dir,
but overlay tools does not support listing redirected dir's childeren,
so it also has that single added directory /renamed_dir.

Signed-off-by: Amir Goldstein <amir73il@gmail.com>

amir73il committed:
Use variables in script

Instead of using full paths in script, use paths relative to $LOWERDIR
and $UPPERDIR variables.

Signed-off-by: Amir Goldstein <amir73il@gmail.com>

amir73il committed:
Implement overlay deref command

copy changes from upperdir to a new upperdir unfolding redirect and metacopy
to get a portable overlay upperdir layer.

Signed-off-by: Amir Goldstein <amir73il@gmail.com>

kmxz commented:
Thanks for the PR! I'll take a look quickly.

kmxz commented on logic.c:
```diff
@@ -202,140 +333,163 @@ int list_deleted_files(const char *path, size_t root_path_len, bool verbose) { /
+    // merging redirects is not supported, we must abort merge so redirected lower (under whiteout) won't be deleted
+    // upper_path may be hiding the directory in lower_path, but there may be another redirect upper pointing at it
+    if (redirect) {
+        fprintf(stderr, "Found redirect on %s. merging redirect is not supported - Abort.\n", upper_path);
```
nit: merging -> Merging?

kmxz commented on main.c:
```diff
@@ -175,15 +216,24 @@ int main(int argc, char *argv[]) {
+            }
+            script = create_shell_script(filename_template);
+            if (script == NULL) { fprintf(stderr, "Script file cannot be created.\n"); return EXIT_FAILURE; }
+	    out = deref(mnt, upper, script);
```
nit: two more spaces?

kmxz commented on sh.c:
```diff
@@ -17,10 +27,28 @@ FILE* create_shell_script(char *tmp_path_buffer) {
     fprintf(f, "# This shell script is generated by overlayfs-tools on %s\n", ctime (&rawtime));
+    for (int i=0; i < NUM_VARS; i++)
+        if (vars[i])
+            fprintf(f, "%s='%s'\n", var_names[i], vars[i]);
```
It may not work well when `vars[i]` contains single quotes.

amir73il replied:
in main.c vars[] is assigned a string returned from realpath(3). It does not contain any quotes, even if quotes are passed in command line

kmxz replied:
I mean the case where the actual directory name itself include quotes.

amir73il replied:
Can you show this with an example?
Actually run overlay that generates a bad script?

kmxz replied:
Here's an example: I set up it with `mkdir "/tmp/test's upper"; mkdir "/tmp/test's lower"; touch "/tmp/test's upper/foo"`, and run `sudo ./overlay merge -l /tmp/test\'s\ lower/ -u /tmp/test\'s\ upper/`.

The generated shell script looks like:

```
LOWERDIR='/tmp/test's lower'
UPPERDIR='/tmp/test's upper'
rm -rf $LOWERDIR'/foo.txt'
```

The the single quote in `test's` breaks the string literal, and the script fails to execute.

amir73il replied:
I see. I am not an expect in quoting as you can see.
So I need to fix create_shell_script() to use quote().
Let me try to fix that.

amir73il replied:
done. please verify I did not miss anything

kmxz reviewed (changes requested):
Thanks for the massive improvements!

I'm not very familiar with the updates in OverlayFS, so I'm not confident in proofreading the logic. But the code surely looks great. I'll definitely merge it.

amir73il committed:
Perform upperdir/lowerdir modifications on a backup copy

For layers modifying scripts 'vaccuum' and 'merge' allow specifying
command line arguments -L <newlower> and -U <newupper>.

If those arguments are specified, the script will make a backup copy of
lowerdir and/or upperdir and will make modifications to the backup copy
instead of modifying the original.

If niether upperdir nor lowerdir are going to be modified, relax the
check for mounted overlayfs.

Signed-off-by: Amir Goldstein <amir73il@gmail.com>

kmxz commented:
Merged. Much appreciation for the great contribution!
